{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from gzip import open as gopen\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R1_count</th>\n",
       "      <th>R2_count</th>\n",
       "      <th>R1_mode_length</th>\n",
       "      <th>R1_count_at_mode_length</th>\n",
       "      <th>R2_mode_length</th>\n",
       "      <th>R2_count_at_mode_length</th>\n",
       "      <th>Mode_overlap</th>\n",
       "      <th>Count_at_mode_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AS3006</th>\n",
       "      <td>2295</td>\n",
       "      <td>3073</td>\n",
       "      <td>130</td>\n",
       "      <td>2295</td>\n",
       "      <td>130</td>\n",
       "      <td>2189</td>\n",
       "      <td>8</td>\n",
       "      <td>1633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3041</th>\n",
       "      <td>9097</td>\n",
       "      <td>9895</td>\n",
       "      <td>130</td>\n",
       "      <td>9097</td>\n",
       "      <td>130</td>\n",
       "      <td>8827</td>\n",
       "      <td>8</td>\n",
       "      <td>6856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3019</th>\n",
       "      <td>11429</td>\n",
       "      <td>11774</td>\n",
       "      <td>130</td>\n",
       "      <td>11429</td>\n",
       "      <td>130</td>\n",
       "      <td>11129</td>\n",
       "      <td>8</td>\n",
       "      <td>8318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3000</th>\n",
       "      <td>19931</td>\n",
       "      <td>20262</td>\n",
       "      <td>130</td>\n",
       "      <td>19931</td>\n",
       "      <td>130</td>\n",
       "      <td>19463</td>\n",
       "      <td>8</td>\n",
       "      <td>15092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3014</th>\n",
       "      <td>14791</td>\n",
       "      <td>15110</td>\n",
       "      <td>130</td>\n",
       "      <td>14791</td>\n",
       "      <td>130</td>\n",
       "      <td>14488</td>\n",
       "      <td>8</td>\n",
       "      <td>11242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3071</th>\n",
       "      <td>9264</td>\n",
       "      <td>9490</td>\n",
       "      <td>130</td>\n",
       "      <td>9264</td>\n",
       "      <td>130</td>\n",
       "      <td>9055</td>\n",
       "      <td>8</td>\n",
       "      <td>7021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3073</th>\n",
       "      <td>4010</td>\n",
       "      <td>5033</td>\n",
       "      <td>130</td>\n",
       "      <td>4010</td>\n",
       "      <td>130</td>\n",
       "      <td>3865</td>\n",
       "      <td>8</td>\n",
       "      <td>3060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3005</th>\n",
       "      <td>14525</td>\n",
       "      <td>15240</td>\n",
       "      <td>130</td>\n",
       "      <td>14525</td>\n",
       "      <td>130</td>\n",
       "      <td>14115</td>\n",
       "      <td>8</td>\n",
       "      <td>10842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3063</th>\n",
       "      <td>11042</td>\n",
       "      <td>11668</td>\n",
       "      <td>130</td>\n",
       "      <td>11042</td>\n",
       "      <td>130</td>\n",
       "      <td>10705</td>\n",
       "      <td>8</td>\n",
       "      <td>8186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3066</th>\n",
       "      <td>4698</td>\n",
       "      <td>5285</td>\n",
       "      <td>130</td>\n",
       "      <td>4698</td>\n",
       "      <td>130</td>\n",
       "      <td>4563</td>\n",
       "      <td>8</td>\n",
       "      <td>3406</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       R1_count R2_count R1_mode_length R1_count_at_mode_length  \\\n",
       "AS3006     2295     3073            130                    2295   \n",
       "AS3041     9097     9895            130                    9097   \n",
       "AS3019    11429    11774            130                   11429   \n",
       "AS3000    19931    20262            130                   19931   \n",
       "AS3014    14791    15110            130                   14791   \n",
       "...         ...      ...            ...                     ...   \n",
       "AS3071     9264     9490            130                    9264   \n",
       "AS3073     4010     5033            130                    4010   \n",
       "AS3005    14525    15240            130                   14525   \n",
       "AS3063    11042    11668            130                   11042   \n",
       "AS3066     4698     5285            130                    4698   \n",
       "\n",
       "       R2_mode_length R2_count_at_mode_length Mode_overlap  \\\n",
       "AS3006            130                    2189            8   \n",
       "AS3041            130                    8827            8   \n",
       "AS3019            130                   11129            8   \n",
       "AS3000            130                   19463            8   \n",
       "AS3014            130                   14488            8   \n",
       "...               ...                     ...          ...   \n",
       "AS3071            130                    9055            8   \n",
       "AS3073            130                    3865            8   \n",
       "AS3005            130                   14115            8   \n",
       "AS3063            130                   10705            8   \n",
       "AS3066            130                    4563            8   \n",
       "\n",
       "       Count_at_mode_overlap  \n",
       "AS3006                  1633  \n",
       "AS3041                  6856  \n",
       "AS3019                  8318  \n",
       "AS3000                 15092  \n",
       "AS3014                 11242  \n",
       "...                      ...  \n",
       "AS3071                  7021  \n",
       "AS3073                  3060  \n",
       "AS3005                 10842  \n",
       "AS3063                  8186  \n",
       "AS3066                  3406  \n",
       "\n",
       "[74 rows x 8 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!mkdir -p cut_reads/16SrRNA\n",
    "\n",
    "for f in glob.glob('raw_data/16SrRNA/AS*_S*_L001_R*_001.fastq.gz'):\n",
    "    outf = 'cut_reads/16SrRNA/%s' % f.split('/')[-1]\n",
    "    smpl = f.split('/')[-1].split('_')[0]\n",
    "    records = SeqIO.parse(gopen(f,'rt'),'fastq')\n",
    "    with gopen(outf,'wt') as hndl:\n",
    "        for r in records:\n",
    "            \n",
    "            # Primer dimers were sequenced. They have monomorphic ends.\n",
    "            # I skip sequences that were close to monomorphic in the last\n",
    "            # 40 base pairs\n",
    "            counts = Counter(str(r[110:150].seq).upper())\n",
    "            proportions = {i: counts[i]/sum(counts.values()) for i in counts.keys()}\n",
    "            if any([proportions[i] > 0.5 for i in proportions.keys()]):\n",
    "                continue\n",
    "                \n",
    "            hndl.write(r[20:150].format('fastq'))\n",
    "\n",
    "!mkdir -p merged/16SrRNA\n",
    "\n",
    "df = pd.DataFrame(columns=['R1_count','R2_count',\n",
    "                           'R1_mode_length','R1_count_at_mode_length',\n",
    "                           'R2_mode_length','R2_count_at_mode_length',\n",
    "                           'Mode_overlap','Count_at_mode_overlap'])\n",
    "\n",
    "for f1 in glob.glob('cut_reads/16SrRNA/AS*_S*_L001_R1_001.fastq.gz'):\n",
    "    f2 = f1.replace('_R1_','_R2_')\n",
    "    \n",
    "    smpl = f1.split('/')[-1].split('_')[0]\n",
    "    \n",
    "    merged = gopen('merged/16SrRNA/%s' % f1.split('/')[-1], 'wt')\n",
    "\n",
    "    r1s = list(SeqIO.parse(gopen(f1,'rt'),'fastq'))\n",
    "    r2s = SeqIO.to_dict(SeqIO.parse(gopen(f2,'rt'),'fastq'))\n",
    "    \n",
    "    count_r1 = len(r1s)\n",
    "    count_r2 = len(r2s)\n",
    "    \n",
    "    overlapps = []\n",
    "    r1_lengths= []\n",
    "    r2_lengths= []\n",
    "    \n",
    "    for r1 in r1s:\n",
    "        r1seq = str(r1.seq)\n",
    "        r1_lengths.append(len(r1))\n",
    "        try:\n",
    "            r2seq = str(r2s[r1.id].seq.reverse_complement())\n",
    "        except:\n",
    "            continue\n",
    "        r2_lengths.append(len(r2seq))\n",
    "        over_len = 30\n",
    "        r1_window = r1seq[-1*over_len:]\n",
    "        r2_window = r2seq[:over_len]\n",
    "        while over_len > 0 and r1_window != r2_window:\n",
    "            over_len -= 1\n",
    "            r1_window = r1seq[-1*over_len:]\n",
    "            r2_window = r2seq[:over_len]\n",
    "        overlapps.append(over_len)\n",
    "        if over_len > 5:\n",
    "            r2 = r2s[r1.id]\n",
    "            m = r1 + r2.reverse_complement()[over_len:]\n",
    "            m.id = r1.id\n",
    "            m.description = r1.description\n",
    "            merged.write(m.format('fastq'))\n",
    "    merged.close()        \n",
    "    if len(overlapps) == 0:\n",
    "        print( smpl)\n",
    "            \n",
    "    ovlp_mode, ovlp_count_at_mode = sorted(Counter(overlapps).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    r1len_mode, r1len_count_at_mode = sorted(Counter(r1_lengths).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    r2len_mode, r2len_count_at_mode = sorted(Counter(r2_lengths).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    df.loc[smpl] = [count_r1,count_r2,r1len_mode,r1len_count_at_mode,r2len_mode,r2len_count_at_mode,ovlp_mode,ovlp_count_at_mode]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('merged_16SrRNA_read_stas.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime tools import \\\n",
    "  --type 'SampleData[SequencesWithQuality]' \\\n",
    "  --input-path merged/16SrRNA \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path merged/16SrRNA/demux.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory â€˜dada/16SrRNAâ€™: File exists\n",
      "\u001b[32mSaved FeatureTable[Frequency] to: dada/16SrRNA/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada/16SrRNA/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: dada/16SrRNA/stats.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!mkdir dada/16SrRNA\n",
    "!qiime dada2 denoise-single \\\n",
    "  --i-demultiplexed-seqs merged/16SrRNA/demux.qza \\\n",
    "  --p-trim-left 0 \\\n",
    "  --p-trunc-len 250 \\\n",
    "  --p-max-ee 5.0 \\\n",
    "  --o-representative-sequences dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-table dada/16SrRNA/table.qza \\\n",
    "  --o-denoising-stats dada/16SrRNA/stats.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[Taxonomy] to: dada/16SrRNA/taxonomy.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier classify-sklearn \\\n",
    "  --i-classifier DB/silva-132-99-515-806-nb-classifier.qza \\\n",
    "  --i-reads dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-classification dada/16SrRNA/taxonomy.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/16SrRNA/aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/16SrRNA/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: dada/16SrRNA/unrooted-tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: dada/16SrRNA/rooted-tree.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-alignment dada/16SrRNA/aligned-rep-seqs.qza \\\n",
    "  --o-masked-alignment dada/16SrRNA/masked-aligned-rep-seqs.qza \\\n",
    "  --o-tree dada/16SrRNA/unrooted-tree.qza \\\n",
    "  --o-rooted-tree dada/16SrRNA/rooted-tree.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
