{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import hashlib\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract qz files for dada2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘DB’: File exists\n",
      "mkdir: cannot create directory ‘temp’: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir DB\n",
    "!mkdir temp\n",
    "!mkdir -p dada/ITS1 \n",
    "\n",
    "for i in glob.glob('raw_data/ITS1/*.fastq.gz'):\n",
    "    outf = 'temp/%s' % i.split('/')[-1].replace('.gz','')\n",
    "    !zcat $i > $outf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write dada2 script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ITS1_DADA2.R\n"
     ]
    }
   ],
   "source": [
    "%%file ITS1_DADA2.R\n",
    "\n",
    "library(Matrix);\n",
    "library(dada2);\n",
    "\n",
    "path <- \"~/GeoDiversity/temp\" # CHANGE ME to the directory containing the fastq files after unzipping.\n",
    "# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq\n",
    "fnFs <- sort(list.files(path, pattern=\"_R1_001.fastq\", full.names = TRUE))\n",
    "fnRs <- sort(list.files(path, pattern=\"_R2_001.fastq\", full.names = TRUE))\n",
    "# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq\n",
    "sample.names <- sapply(strsplit(basename(fnFs), \"_\"), `[`, 1)\n",
    "# Place filtered files in filtered/ subdirectory\n",
    "path1 = \"~/GeoDiversity/cut_reads\"                                                        \n",
    "filtFs <- file.path(path1, \"ITS1\", paste0(sample.names, \"_F_filt.fastq.gz\"))\n",
    "filtRs <- file.path(path1, \"ITS1\", paste0(sample.names, \"_R_filt.fastq.gz\"))                                                        \n",
    "out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(125,125),\n",
    "              maxN=0, maxEE=c(4,5), truncQ=10, rm.phix=TRUE, trimLeft=0,\n",
    "              compress=TRUE, multithread=TRUE)\n",
    "# learn F errors\n",
    "errF <- learnErrors(filtFs, multithread=TRUE)\n",
    "print('learn R errors')\n",
    "errR <- learnErrors(filtRs, multithread=TRUE)\n",
    "\n",
    "# derep F\n",
    "derepFs <- derepFastq(filtFs, verbose=TRUE)\n",
    "# derep R\n",
    "derepRs <- derepFastq(filtRs, verbose=TRUE)\n",
    "# Name the derep-class objects by the sample names\n",
    "names(derepFs) <- sample.names\n",
    "names(derepRs) <- sample.names\n",
    "\n",
    "# dada F\n",
    "dadaFs <- dada(derepFs, err=errF, multithread=TRUE)\n",
    "# dada R\n",
    "dadaRs <- dada(derepRs, err=errR, multithread=TRUE)\n",
    "\n",
    "# merging\n",
    "mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE, justConcatenate=TRUE)\n",
    "seqtab <- makeSequenceTable(mergers)\n",
    "\n",
    "# chimera removal                                                    \n",
    "seqtab.nochim <- removeBimeraDenovo(seqtab, method=\"consensus\", multithread=TRUE, verbose=TRUE)\n",
    "\n",
    "# assign taxonomy                                                        \n",
    "taxa <- assignTaxonomy(seqtab, \"~/GeoDiversity/DB/sh_general_release_dynamic_02.02.2019.fasta\")\n",
    "\n",
    "# write outputs                                                   \n",
    "uniquesToFasta(seqtab, 'dada/ITS1/rep-seqs.fasta')\n",
    "write.csv(seqtab, file = \"dada/ITS1/table.csv\")\n",
    "write.csv(taxa, file = \"dada/ITS1/taxa.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run dada2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: Rcpp\n",
      "Creating output directory: /home/amir/GeoDiversity/cut_reads/ITS1\n",
      "101497500 total bases in 811980 reads from 62 samples will be used for learning the error rates.\n",
      "[1] \"learn R errors\"\n",
      "101497500 total bases in 811980 reads from 62 samples will be used for learning the error rates.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3000_F_filt.fastq.gz\n",
      "Encountered 5657 unique sequences from 20211 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3001_F_filt.fastq.gz\n",
      "Encountered 3698 unique sequences from 15477 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3002_F_filt.fastq.gz\n",
      "Encountered 3981 unique sequences from 17678 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3003_F_filt.fastq.gz\n",
      "Encountered 1817 unique sequences from 8343 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3004_F_filt.fastq.gz\n",
      "Encountered 3417 unique sequences from 12793 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3005_F_filt.fastq.gz\n",
      "Encountered 2775 unique sequences from 13776 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3006_F_filt.fastq.gz\n",
      "Encountered 1296 unique sequences from 5036 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3007_F_filt.fastq.gz\n",
      "Encountered 3213 unique sequences from 16296 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3008_F_filt.fastq.gz\n",
      "Encountered 5133 unique sequences from 17768 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3009_F_filt.fastq.gz\n",
      "Encountered 3984 unique sequences from 16561 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3010_F_filt.fastq.gz\n",
      "Encountered 4277 unique sequences from 16722 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3011_F_filt.fastq.gz\n",
      "Encountered 2620 unique sequences from 9639 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3012_F_filt.fastq.gz\n",
      "Encountered 3754 unique sequences from 15861 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3013_F_filt.fastq.gz\n",
      "Encountered 1674 unique sequences from 7320 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3014_F_filt.fastq.gz\n",
      "Encountered 3214 unique sequences from 14952 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3015_F_filt.fastq.gz\n",
      "Encountered 3834 unique sequences from 15158 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3016_F_filt.fastq.gz\n",
      "Encountered 1885 unique sequences from 5759 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3017_F_filt.fastq.gz\n",
      "Encountered 3058 unique sequences from 11935 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3018_F_filt.fastq.gz\n",
      "Encountered 3733 unique sequences from 12424 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3019_F_filt.fastq.gz\n",
      "Encountered 2559 unique sequences from 9286 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3020_F_filt.fastq.gz\n",
      "Encountered 3330 unique sequences from 13108 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3021_F_filt.fastq.gz\n",
      "Encountered 3103 unique sequences from 10514 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3022_F_filt.fastq.gz\n",
      "Encountered 3219 unique sequences from 12929 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3023_F_filt.fastq.gz\n",
      "Encountered 3979 unique sequences from 15853 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3024_F_filt.fastq.gz\n",
      "Encountered 344 unique sequences from 807 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3025_F_filt.fastq.gz\n",
      "Encountered 3714 unique sequences from 15244 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3026_F_filt.fastq.gz\n",
      "Encountered 3071 unique sequences from 11936 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3027_F_filt.fastq.gz\n",
      "Encountered 2483 unique sequences from 11276 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3028_F_filt.fastq.gz\n",
      "Encountered 3392 unique sequences from 16128 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3029_F_filt.fastq.gz\n",
      "Encountered 3655 unique sequences from 15992 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3030_F_filt.fastq.gz\n",
      "Encountered 1 unique sequences from 1 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3031_F_filt.fastq.gz\n",
      "Encountered 5092 unique sequences from 18476 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3032_F_filt.fastq.gz\n",
      "Encountered 2432 unique sequences from 9439 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3033_F_filt.fastq.gz\n",
      "Encountered 3789 unique sequences from 12731 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3034_F_filt.fastq.gz\n",
      "Encountered 3169 unique sequences from 12757 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3035_F_filt.fastq.gz\n",
      "Encountered 1305 unique sequences from 3422 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3036_F_filt.fastq.gz\n",
      "Encountered 2813 unique sequences from 9084 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3037_F_filt.fastq.gz\n",
      "Encountered 3127 unique sequences from 11727 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3038_F_filt.fastq.gz\n",
      "Encountered 3502 unique sequences from 16714 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3039_F_filt.fastq.gz\n",
      "Encountered 5103 unique sequences from 18978 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3040_F_filt.fastq.gz\n",
      "Encountered 2456 unique sequences from 10798 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3041_F_filt.fastq.gz\n",
      "Encountered 3440 unique sequences from 11674 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3042_F_filt.fastq.gz\n",
      "Encountered 5020 unique sequences from 15936 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3043_F_filt.fastq.gz\n",
      "Encountered 3039 unique sequences from 21043 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3044_F_filt.fastq.gz\n",
      "Encountered 3041 unique sequences from 10830 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3045_F_filt.fastq.gz\n",
      "Encountered 2718 unique sequences from 9023 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3046_F_filt.fastq.gz\n",
      "Encountered 4478 unique sequences from 14647 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3047_F_filt.fastq.gz\n",
      "Encountered 5648 unique sequences from 20329 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3048_F_filt.fastq.gz\n",
      "Encountered 2705 unique sequences from 14795 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3049_F_filt.fastq.gz\n",
      "Encountered 3211 unique sequences from 12573 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3050_F_filt.fastq.gz\n",
      "Encountered 2033 unique sequences from 14622 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3051_F_filt.fastq.gz\n",
      "Encountered 2043 unique sequences from 7052 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3052_F_filt.fastq.gz\n",
      "Encountered 4285 unique sequences from 14043 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3053_F_filt.fastq.gz\n",
      "Encountered 3463 unique sequences from 12157 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3054_F_filt.fastq.gz\n",
      "Encountered 3641 unique sequences from 12957 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3055_F_filt.fastq.gz\n",
      "Encountered 4869 unique sequences from 18857 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3056_F_filt.fastq.gz\n",
      "Encountered 4222 unique sequences from 16002 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3057_F_filt.fastq.gz\n",
      "Encountered 4880 unique sequences from 21108 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3058_F_filt.fastq.gz\n",
      "Encountered 4608 unique sequences from 18416 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3059_F_filt.fastq.gz\n",
      "Encountered 2575 unique sequences from 10824 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3060_F_filt.fastq.gz\n",
      "Encountered 2118 unique sequences from 7915 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3061_F_filt.fastq.gz\n",
      "Encountered 3432 unique sequences from 16268 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3062_F_filt.fastq.gz\n",
      "Encountered 170 unique sequences from 388 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3063_F_filt.fastq.gz\n",
      "Encountered 4419 unique sequences from 24685 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3064_F_filt.fastq.gz\n",
      "Encountered 4146 unique sequences from 17715 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3065_F_filt.fastq.gz\n",
      "Encountered 3521 unique sequences from 13058 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3066_F_filt.fastq.gz\n",
      "Encountered 4762 unique sequences from 18292 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3067_F_filt.fastq.gz\n",
      "Encountered 3796 unique sequences from 14336 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3070_F_filt.fastq.gz\n",
      "Encountered 3682 unique sequences from 10941 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3071_F_filt.fastq.gz\n",
      "Encountered 3910 unique sequences from 16315 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3072_F_filt.fastq.gz\n",
      "Encountered 3552 unique sequences from 17844 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3073_F_filt.fastq.gz\n",
      "Encountered 866 unique sequences from 2878 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3074_F_filt.fastq.gz\n",
      "Encountered 473 unique sequences from 1253 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3075_F_filt.fastq.gz\n",
      "Encountered 554 unique sequences from 1962 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3000_R_filt.fastq.gz\n",
      "Encountered 8423 unique sequences from 20211 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3001_R_filt.fastq.gz\n",
      "Encountered 5655 unique sequences from 15477 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3002_R_filt.fastq.gz\n",
      "Encountered 6484 unique sequences from 17678 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3003_R_filt.fastq.gz\n",
      "Encountered 2588 unique sequences from 8343 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3004_R_filt.fastq.gz\n",
      "Encountered 5185 unique sequences from 12793 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3005_R_filt.fastq.gz\n",
      "Encountered 4376 unique sequences from 13776 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3006_R_filt.fastq.gz\n",
      "Encountered 2260 unique sequences from 5036 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3007_R_filt.fastq.gz\n",
      "Encountered 4795 unique sequences from 16296 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3008_R_filt.fastq.gz\n",
      "Encountered 7777 unique sequences from 17768 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3009_R_filt.fastq.gz\n",
      "Encountered 6543 unique sequences from 16561 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3010_R_filt.fastq.gz\n",
      "Encountered 6287 unique sequences from 16722 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3011_R_filt.fastq.gz\n",
      "Encountered 4077 unique sequences from 9639 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3012_R_filt.fastq.gz\n",
      "Encountered 5776 unique sequences from 15861 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3013_R_filt.fastq.gz\n",
      "Encountered 2407 unique sequences from 7320 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3014_R_filt.fastq.gz\n",
      "Encountered 5274 unique sequences from 14952 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3015_R_filt.fastq.gz\n",
      "Encountered 6090 unique sequences from 15158 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3016_R_filt.fastq.gz\n",
      "Encountered 3010 unique sequences from 5759 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3017_R_filt.fastq.gz\n",
      "Encountered 4672 unique sequences from 11935 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3018_R_filt.fastq.gz\n",
      "Encountered 5301 unique sequences from 12424 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3019_R_filt.fastq.gz\n",
      "Encountered 3878 unique sequences from 9286 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3020_R_filt.fastq.gz\n",
      "Encountered 5221 unique sequences from 13108 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3021_R_filt.fastq.gz\n",
      "Encountered 4536 unique sequences from 10514 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3022_R_filt.fastq.gz\n",
      "Encountered 4886 unique sequences from 12929 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3023_R_filt.fastq.gz\n",
      "Encountered 6504 unique sequences from 15853 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3024_R_filt.fastq.gz\n",
      "Encountered 460 unique sequences from 807 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3025_R_filt.fastq.gz\n",
      "Encountered 5481 unique sequences from 15244 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3026_R_filt.fastq.gz\n",
      "Encountered 4612 unique sequences from 11936 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3027_R_filt.fastq.gz\n",
      "Encountered 3769 unique sequences from 11276 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3028_R_filt.fastq.gz\n",
      "Encountered 6121 unique sequences from 16128 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3029_R_filt.fastq.gz\n",
      "Encountered 5754 unique sequences from 15992 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3030_R_filt.fastq.gz\n",
      "Encountered 1 unique sequences from 1 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3031_R_filt.fastq.gz\n",
      "Encountered 7230 unique sequences from 18476 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3032_R_filt.fastq.gz\n",
      "Encountered 3686 unique sequences from 9439 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3033_R_filt.fastq.gz\n",
      "Encountered 5778 unique sequences from 12731 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3034_R_filt.fastq.gz\n",
      "Encountered 4804 unique sequences from 12757 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3035_R_filt.fastq.gz\n",
      "Encountered 1879 unique sequences from 3422 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3036_R_filt.fastq.gz\n",
      "Encountered 3726 unique sequences from 9084 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3037_R_filt.fastq.gz\n",
      "Encountered 4531 unique sequences from 11727 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3038_R_filt.fastq.gz\n",
      "Encountered 5895 unique sequences from 16714 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3039_R_filt.fastq.gz\n",
      "Encountered 7112 unique sequences from 18978 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3040_R_filt.fastq.gz\n",
      "Encountered 3575 unique sequences from 10798 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3041_R_filt.fastq.gz\n",
      "Encountered 5041 unique sequences from 11674 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3042_R_filt.fastq.gz\n",
      "Encountered 7653 unique sequences from 15936 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3043_R_filt.fastq.gz\n",
      "Encountered 4832 unique sequences from 21043 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3044_R_filt.fastq.gz\n",
      "Encountered 4461 unique sequences from 10830 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3045_R_filt.fastq.gz\n",
      "Encountered 3848 unique sequences from 9023 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3046_R_filt.fastq.gz\n",
      "Encountered 6555 unique sequences from 14647 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3047_R_filt.fastq.gz\n",
      "Encountered 8275 unique sequences from 20329 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3048_R_filt.fastq.gz\n",
      "Encountered 4426 unique sequences from 14795 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3049_R_filt.fastq.gz\n",
      "Encountered 4651 unique sequences from 12573 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3050_R_filt.fastq.gz\n",
      "Encountered 3525 unique sequences from 14622 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3051_R_filt.fastq.gz\n",
      "Encountered 3151 unique sequences from 7052 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3052_R_filt.fastq.gz\n",
      "Encountered 6428 unique sequences from 14043 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3053_R_filt.fastq.gz\n",
      "Encountered 5054 unique sequences from 12157 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3054_R_filt.fastq.gz\n",
      "Encountered 5596 unique sequences from 12957 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3055_R_filt.fastq.gz\n",
      "Encountered 7631 unique sequences from 18857 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3056_R_filt.fastq.gz\n",
      "Encountered 6466 unique sequences from 16002 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3057_R_filt.fastq.gz\n",
      "Encountered 7878 unique sequences from 21108 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3058_R_filt.fastq.gz\n",
      "Encountered 7681 unique sequences from 18416 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3059_R_filt.fastq.gz\n",
      "Encountered 3512 unique sequences from 10824 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3060_R_filt.fastq.gz\n",
      "Encountered 3013 unique sequences from 7915 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3061_R_filt.fastq.gz\n",
      "Encountered 5119 unique sequences from 16268 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3062_R_filt.fastq.gz\n",
      "Encountered 241 unique sequences from 388 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3063_R_filt.fastq.gz\n",
      "Encountered 6363 unique sequences from 24685 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3064_R_filt.fastq.gz\n",
      "Encountered 6808 unique sequences from 17715 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3065_R_filt.fastq.gz\n",
      "Encountered 5073 unique sequences from 13058 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3066_R_filt.fastq.gz\n",
      "Encountered 6794 unique sequences from 18292 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3067_R_filt.fastq.gz\n",
      "Encountered 5594 unique sequences from 14336 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3070_R_filt.fastq.gz\n",
      "Encountered 5462 unique sequences from 10941 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3071_R_filt.fastq.gz\n",
      "Encountered 5656 unique sequences from 16315 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3072_R_filt.fastq.gz\n",
      "Encountered 5749 unique sequences from 17844 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3073_R_filt.fastq.gz\n",
      "Encountered 1238 unique sequences from 2878 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3074_R_filt.fastq.gz\n",
      "Encountered 751 unique sequences from 1253 total sequences read.\n",
      "Dereplicating sequence entries in Fastq file: ~/GeoDiversity/cut_reads/ITS1/ITS-AS3075_R_filt.fastq.gz\n",
      "Encountered 835 unique sequences from 1962 total sequences read.\n",
      "Sample 1 - 20211 reads in 5657 unique sequences.\n",
      "Sample 2 - 15477 reads in 3698 unique sequences.\n",
      "Sample 3 - 17678 reads in 3981 unique sequences.\n",
      "Sample 4 - 8343 reads in 1817 unique sequences.\n",
      "Sample 5 - 12793 reads in 3417 unique sequences.\n",
      "Sample 6 - 13776 reads in 2775 unique sequences.\n",
      "Sample 7 - 5036 reads in 1296 unique sequences.\n",
      "Sample 8 - 16296 reads in 3213 unique sequences.\n",
      "Sample 9 - 17768 reads in 5133 unique sequences.\n",
      "Sample 10 - 16561 reads in 3984 unique sequences.\n",
      "Sample 11 - 16722 reads in 4277 unique sequences.\n",
      "Sample 12 - 9639 reads in 2620 unique sequences.\n",
      "Sample 13 - 15861 reads in 3754 unique sequences.\n",
      "Sample 14 - 7320 reads in 1674 unique sequences.\n",
      "Sample 15 - 14952 reads in 3214 unique sequences.\n",
      "Sample 16 - 15158 reads in 3834 unique sequences.\n",
      "Sample 17 - 5759 reads in 1885 unique sequences.\n",
      "Sample 18 - 11935 reads in 3058 unique sequences.\n",
      "Sample 19 - 12424 reads in 3733 unique sequences.\n",
      "Sample 20 - 9286 reads in 2559 unique sequences.\n",
      "Sample 21 - 13108 reads in 3330 unique sequences.\n",
      "Sample 22 - 10514 reads in 3103 unique sequences.\n",
      "Sample 23 - 12929 reads in 3219 unique sequences.\n",
      "Sample 24 - 15853 reads in 3979 unique sequences.\n",
      "Sample 25 - 807 reads in 344 unique sequences.\n",
      "Sample 26 - 15244 reads in 3714 unique sequences.\n",
      "Sample 27 - 11936 reads in 3071 unique sequences.\n",
      "Sample 28 - 11276 reads in 2483 unique sequences.\n",
      "Sample 29 - 16128 reads in 3392 unique sequences.\n",
      "Sample 30 - 15992 reads in 3655 unique sequences.\n",
      "Sample 31 - 1 reads in 1 unique sequences.\n",
      "Sample 32 - 18476 reads in 5092 unique sequences.\n",
      "Sample 33 - 9439 reads in 2432 unique sequences.\n",
      "Sample 34 - 12731 reads in 3789 unique sequences.\n",
      "Sample 35 - 12757 reads in 3169 unique sequences.\n",
      "Sample 36 - 3422 reads in 1305 unique sequences.\n",
      "Sample 37 - 9084 reads in 2813 unique sequences.\n",
      "Sample 38 - 11727 reads in 3127 unique sequences.\n",
      "Sample 39 - 16714 reads in 3502 unique sequences.\n",
      "Sample 40 - 18978 reads in 5103 unique sequences.\n",
      "Sample 41 - 10798 reads in 2456 unique sequences.\n",
      "Sample 42 - 11674 reads in 3440 unique sequences.\n",
      "Sample 43 - 15936 reads in 5020 unique sequences.\n",
      "Sample 44 - 21043 reads in 3039 unique sequences.\n",
      "Sample 45 - 10830 reads in 3041 unique sequences.\n",
      "Sample 46 - 9023 reads in 2718 unique sequences.\n",
      "Sample 47 - 14647 reads in 4478 unique sequences.\n",
      "Sample 48 - 20329 reads in 5648 unique sequences.\n",
      "Sample 49 - 14795 reads in 2705 unique sequences.\n",
      "Sample 50 - 12573 reads in 3211 unique sequences.\n",
      "Sample 51 - 14622 reads in 2033 unique sequences.\n",
      "Sample 52 - 7052 reads in 2043 unique sequences.\n",
      "Sample 53 - 14043 reads in 4285 unique sequences.\n",
      "Sample 54 - 12157 reads in 3463 unique sequences.\n",
      "Sample 55 - 12957 reads in 3641 unique sequences.\n",
      "Sample 56 - 18857 reads in 4869 unique sequences.\n",
      "Sample 57 - 16002 reads in 4222 unique sequences.\n",
      "Sample 58 - 21108 reads in 4880 unique sequences.\n",
      "Sample 59 - 18416 reads in 4608 unique sequences.\n",
      "Sample 60 - 10824 reads in 2575 unique sequences.\n",
      "Sample 61 - 7915 reads in 2118 unique sequences.\n",
      "Sample 62 - 16268 reads in 3432 unique sequences.\n",
      "Sample 63 - 388 reads in 170 unique sequences.\n",
      "Sample 64 - 24685 reads in 4419 unique sequences.\n",
      "Sample 65 - 17715 reads in 4146 unique sequences.\n",
      "Sample 66 - 13058 reads in 3521 unique sequences.\n",
      "Sample 67 - 18292 reads in 4762 unique sequences.\n",
      "Sample 68 - 14336 reads in 3796 unique sequences.\n",
      "Sample 69 - 10941 reads in 3682 unique sequences.\n",
      "Sample 70 - 16315 reads in 3910 unique sequences.\n",
      "Sample 71 - 17844 reads in 3552 unique sequences.\n",
      "Sample 72 - 2878 reads in 866 unique sequences.\n",
      "Sample 73 - 1253 reads in 473 unique sequences.\n",
      "Sample 74 - 1962 reads in 554 unique sequences.\n",
      "Sample 1 - 20211 reads in 8423 unique sequences.\n",
      "Sample 2 - 15477 reads in 5655 unique sequences.\n",
      "Sample 3 - 17678 reads in 6484 unique sequences.\n",
      "Sample 4 - 8343 reads in 2588 unique sequences.\n",
      "Sample 5 - 12793 reads in 5185 unique sequences.\n",
      "Sample 6 - 13776 reads in 4376 unique sequences.\n",
      "Sample 7 - 5036 reads in 2260 unique sequences.\n",
      "Sample 8 - 16296 reads in 4795 unique sequences.\n",
      "Sample 9 - 17768 reads in 7777 unique sequences.\n",
      "Sample 10 - 16561 reads in 6543 unique sequences.\n",
      "Sample 11 - 16722 reads in 6287 unique sequences.\n",
      "Sample 12 - 9639 reads in 4077 unique sequences.\n",
      "Sample 13 - 15861 reads in 5776 unique sequences.\n",
      "Sample 14 - 7320 reads in 2407 unique sequences.\n",
      "Sample 15 - 14952 reads in 5274 unique sequences.\n",
      "Sample 16 - 15158 reads in 6090 unique sequences.\n",
      "Sample 17 - 5759 reads in 3010 unique sequences.\n",
      "Sample 18 - 11935 reads in 4672 unique sequences.\n",
      "Sample 19 - 12424 reads in 5301 unique sequences.\n",
      "Sample 20 - 9286 reads in 3878 unique sequences.\n",
      "Sample 21 - 13108 reads in 5221 unique sequences.\n",
      "Sample 22 - 10514 reads in 4536 unique sequences.\n",
      "Sample 23 - 12929 reads in 4886 unique sequences.\n",
      "Sample 24 - 15853 reads in 6504 unique sequences.\n",
      "Sample 25 - 807 reads in 460 unique sequences.\n",
      "Sample 26 - 15244 reads in 5481 unique sequences.\n",
      "Sample 27 - 11936 reads in 4612 unique sequences.\n",
      "Sample 28 - 11276 reads in 3769 unique sequences.\n",
      "Sample 29 - 16128 reads in 6121 unique sequences.\n",
      "Sample 30 - 15992 reads in 5754 unique sequences.\n",
      "Sample 31 - 1 reads in 1 unique sequences.\n",
      "Sample 32 - 18476 reads in 7230 unique sequences.\n",
      "Sample 33 - 9439 reads in 3686 unique sequences.\n",
      "Sample 34 - 12731 reads in 5778 unique sequences.\n",
      "Sample 35 - 12757 reads in 4804 unique sequences.\n",
      "Sample 36 - 3422 reads in 1879 unique sequences.\n",
      "Sample 37 - 9084 reads in 3726 unique sequences.\n",
      "Sample 38 - 11727 reads in 4531 unique sequences.\n",
      "Sample 39 - 16714 reads in 5895 unique sequences.\n",
      "Sample 40 - 18978 reads in 7112 unique sequences.\n",
      "Sample 41 - 10798 reads in 3575 unique sequences.\n",
      "Sample 42 - 11674 reads in 5041 unique sequences.\n",
      "Sample 43 - 15936 reads in 7653 unique sequences.\n",
      "Sample 44 - 21043 reads in 4832 unique sequences.\n",
      "Sample 45 - 10830 reads in 4461 unique sequences.\n",
      "Sample 46 - 9023 reads in 3848 unique sequences.\n",
      "Sample 47 - 14647 reads in 6555 unique sequences.\n",
      "Sample 48 - 20329 reads in 8275 unique sequences.\n",
      "Sample 49 - 14795 reads in 4426 unique sequences.\n",
      "Sample 50 - 12573 reads in 4651 unique sequences.\n",
      "Sample 51 - 14622 reads in 3525 unique sequences.\n",
      "Sample 52 - 7052 reads in 3151 unique sequences.\n",
      "Sample 53 - 14043 reads in 6428 unique sequences.\n",
      "Sample 54 - 12157 reads in 5054 unique sequences.\n",
      "Sample 55 - 12957 reads in 5596 unique sequences.\n",
      "Sample 56 - 18857 reads in 7631 unique sequences.\n",
      "Sample 57 - 16002 reads in 6466 unique sequences.\n",
      "Sample 58 - 21108 reads in 7878 unique sequences.\n",
      "Sample 59 - 18416 reads in 7681 unique sequences.\n",
      "Sample 60 - 10824 reads in 3512 unique sequences.\n",
      "Sample 61 - 7915 reads in 3013 unique sequences.\n",
      "Sample 62 - 16268 reads in 5119 unique sequences.\n",
      "Sample 63 - 388 reads in 241 unique sequences.\n",
      "Sample 64 - 24685 reads in 6363 unique sequences.\n",
      "Sample 65 - 17715 reads in 6808 unique sequences.\n",
      "Sample 66 - 13058 reads in 5073 unique sequences.\n",
      "Sample 67 - 18292 reads in 6794 unique sequences.\n",
      "Sample 68 - 14336 reads in 5594 unique sequences.\n",
      "Sample 69 - 10941 reads in 5462 unique sequences.\n",
      "Sample 70 - 16315 reads in 5656 unique sequences.\n",
      "Sample 71 - 17844 reads in 5749 unique sequences.\n",
      "Sample 72 - 2878 reads in 1238 unique sequences.\n",
      "Sample 73 - 1253 reads in 751 unique sequences.\n",
      "Sample 74 - 1962 reads in 835 unique sequences.\n",
      "19088 paired-reads (in 549 unique pairings) successfully merged out of 19088 (in 549 pairings) input.\n",
      "14705 paired-reads (in 318 unique pairings) successfully merged out of 14705 (in 318 pairings) input.\n",
      "16645 paired-reads (in 379 unique pairings) successfully merged out of 16645 (in 379 pairings) input.\n",
      "7957 paired-reads (in 168 unique pairings) successfully merged out of 7957 (in 168 pairings) input.\n",
      "11915 paired-reads (in 271 unique pairings) successfully merged out of 11915 (in 271 pairings) input.\n",
      "13135 paired-reads (in 202 unique pairings) successfully merged out of 13135 (in 202 pairings) input.\n",
      "4741 paired-reads (in 107 unique pairings) successfully merged out of 4741 (in 107 pairings) input.\n",
      "15569 paired-reads (in 189 unique pairings) successfully merged out of 15569 (in 189 pairings) input.\n",
      "16579 paired-reads (in 406 unique pairings) successfully merged out of 16579 (in 406 pairings) input.\n",
      "15652 paired-reads (in 366 unique pairings) successfully merged out of 15652 (in 366 pairings) input.\n",
      "15804 paired-reads (in 450 unique pairings) successfully merged out of 15804 (in 450 pairings) input.\n",
      "9040 paired-reads (in 239 unique pairings) successfully merged out of 9040 (in 239 pairings) input.\n",
      "14942 paired-reads (in 342 unique pairings) successfully merged out of 14942 (in 342 pairings) input.\n",
      "6929 paired-reads (in 139 unique pairings) successfully merged out of 6929 (in 139 pairings) input.\n",
      "14173 paired-reads (in 292 unique pairings) successfully merged out of 14173 (in 292 pairings) input.\n",
      "14393 paired-reads (in 343 unique pairings) successfully merged out of 14393 (in 343 pairings) input.\n",
      "5273 paired-reads (in 165 unique pairings) successfully merged out of 5273 (in 165 pairings) input.\n",
      "11192 paired-reads (in 226 unique pairings) successfully merged out of 11192 (in 226 pairings) input.\n",
      "11645 paired-reads (in 318 unique pairings) successfully merged out of 11645 (in 318 pairings) input.\n",
      "8666 paired-reads (in 262 unique pairings) successfully merged out of 8666 (in 262 pairings) input.\n",
      "12217 paired-reads (in 393 unique pairings) successfully merged out of 12217 (in 393 pairings) input.\n",
      "9837 paired-reads (in 317 unique pairings) successfully merged out of 9837 (in 317 pairings) input.\n",
      "12242 paired-reads (in 242 unique pairings) successfully merged out of 12242 (in 242 pairings) input.\n",
      "15116 paired-reads (in 229 unique pairings) successfully merged out of 15116 (in 229 pairings) input.\n",
      "665 paired-reads (in 30 unique pairings) successfully merged out of 665 (in 30 pairings) input.\n",
      "14355 paired-reads (in 336 unique pairings) successfully merged out of 14355 (in 336 pairings) input.\n",
      "11088 paired-reads (in 341 unique pairings) successfully merged out of 11088 (in 341 pairings) input.\n",
      "10742 paired-reads (in 271 unique pairings) successfully merged out of 10742 (in 271 pairings) input.\n",
      "15207 paired-reads (in 511 unique pairings) successfully merged out of 15207 (in 511 pairings) input.\n",
      "15171 paired-reads (in 302 unique pairings) successfully merged out of 15171 (in 302 pairings) input.\n",
      "1 paired-reads (in 1 unique pairings) successfully merged out of 1 (in 1 pairings) input.\n",
      "17363 paired-reads (in 512 unique pairings) successfully merged out of 17363 (in 512 pairings) input.\n",
      "8859 paired-reads (in 183 unique pairings) successfully merged out of 8859 (in 183 pairings) input.\n",
      "11579 paired-reads (in 361 unique pairings) successfully merged out of 11579 (in 361 pairings) input.\n",
      "12080 paired-reads (in 265 unique pairings) successfully merged out of 12080 (in 265 pairings) input.\n",
      "3061 paired-reads (in 115 unique pairings) successfully merged out of 3061 (in 115 pairings) input.\n",
      "8402 paired-reads (in 321 unique pairings) successfully merged out of 8402 (in 321 pairings) input.\n",
      "10995 paired-reads (in 253 unique pairings) successfully merged out of 10995 (in 253 pairings) input.\n",
      "15855 paired-reads (in 307 unique pairings) successfully merged out of 15855 (in 307 pairings) input.\n",
      "17958 paired-reads (in 579 unique pairings) successfully merged out of 17958 (in 579 pairings) input.\n",
      "10027 paired-reads (in 262 unique pairings) successfully merged out of 10027 (in 262 pairings) input.\n",
      "10745 paired-reads (in 403 unique pairings) successfully merged out of 10745 (in 403 pairings) input.\n",
      "14527 paired-reads (in 397 unique pairings) successfully merged out of 14527 (in 397 pairings) input.\n",
      "20342 paired-reads (in 183 unique pairings) successfully merged out of 20342 (in 183 pairings) input.\n",
      "10043 paired-reads (in 240 unique pairings) successfully merged out of 10043 (in 240 pairings) input.\n",
      "8481 paired-reads (in 208 unique pairings) successfully merged out of 8481 (in 208 pairings) input.\n",
      "13396 paired-reads (in 442 unique pairings) successfully merged out of 13396 (in 442 pairings) input.\n",
      "19238 paired-reads (in 369 unique pairings) successfully merged out of 19238 (in 369 pairings) input.\n",
      "14324 paired-reads (in 150 unique pairings) successfully merged out of 14324 (in 150 pairings) input.\n",
      "11802 paired-reads (in 344 unique pairings) successfully merged out of 11802 (in 344 pairings) input.\n",
      "14190 paired-reads (in 112 unique pairings) successfully merged out of 14190 (in 112 pairings) input.\n",
      "6463 paired-reads (in 225 unique pairings) successfully merged out of 6463 (in 225 pairings) input.\n",
      "12878 paired-reads (in 403 unique pairings) successfully merged out of 12878 (in 403 pairings) input.\n",
      "11094 paired-reads (in 336 unique pairings) successfully merged out of 11094 (in 336 pairings) input.\n",
      "12073 paired-reads (in 384 unique pairings) successfully merged out of 12073 (in 384 pairings) input.\n",
      "17701 paired-reads (in 511 unique pairings) successfully merged out of 17701 (in 511 pairings) input.\n",
      "14918 paired-reads (in 439 unique pairings) successfully merged out of 14918 (in 439 pairings) input.\n",
      "20025 paired-reads (in 544 unique pairings) successfully merged out of 20025 (in 544 pairings) input.\n",
      "17370 paired-reads (in 471 unique pairings) successfully merged out of 17370 (in 471 pairings) input.\n",
      "10318 paired-reads (in 319 unique pairings) successfully merged out of 10318 (in 319 pairings) input.\n",
      "7322 paired-reads (in 259 unique pairings) successfully merged out of 7322 (in 259 pairings) input.\n",
      "15586 paired-reads (in 313 unique pairings) successfully merged out of 15586 (in 313 pairings) input.\n",
      "295 paired-reads (in 19 unique pairings) successfully merged out of 295 (in 19 pairings) input.\n",
      "23945 paired-reads (in 486 unique pairings) successfully merged out of 23945 (in 486 pairings) input.\n",
      "16771 paired-reads (in 417 unique pairings) successfully merged out of 16771 (in 417 pairings) input.\n",
      "12107 paired-reads (in 302 unique pairings) successfully merged out of 12107 (in 302 pairings) input.\n",
      "17224 paired-reads (in 463 unique pairings) successfully merged out of 17224 (in 463 pairings) input.\n",
      "13401 paired-reads (in 480 unique pairings) successfully merged out of 13401 (in 480 pairings) input.\n",
      "9962 paired-reads (in 359 unique pairings) successfully merged out of 9962 (in 359 pairings) input.\n",
      "15473 paired-reads (in 373 unique pairings) successfully merged out of 15473 (in 373 pairings) input.\n",
      "17104 paired-reads (in 257 unique pairings) successfully merged out of 17104 (in 257 pairings) input.\n",
      "2571 paired-reads (in 66 unique pairings) successfully merged out of 2571 (in 66 pairings) input.\n",
      "1007 paired-reads (in 33 unique pairings) successfully merged out of 1007 (in 33 pairings) input.\n",
      "1740 paired-reads (in 63 unique pairings) successfully merged out of 1740 (in 63 pairings) input.\n",
      "Identified 6435 bimeras out of 13224 input sequences.\n",
      "UNITE fungal taxonomic reference detected.\n"
     ]
    }
   ],
   "source": [
    "!Rscript ITS1_DADA2.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import to qiime2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported dada/ITS1/table.biom as BIOMV210Format to dada/ITS1/table.qza\u001b[0m\n",
      "\u001b[32mImported temp1 as DNASequencesDirectoryFormat to dada/ITS1/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mImported temp.tsv as HeaderlessTSVTaxonomyFormat to dada/ITS1/taxonomy.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# add md5 hashes to table\n",
    "df = pd.read_csv('dada/ITS1/table.csv',index_col=0).transpose()\n",
    "df.index = [hashlib.md5(i.encode()).hexdigest() for i in df.index]\n",
    "df.index.name = '#Feature_id'\n",
    "df.to_csv('temp1',sep='\\t')\n",
    "\n",
    "# convert table to biom\n",
    "!biom convert -i temp1 -o dada/ITS1/table.biom --table-type=\"OTU table\" --to-hdf5\n",
    "!rm temp1\n",
    "\n",
    "# convert biom table to qza\n",
    "!qiime tools import \\\n",
    "  --input-path dada/ITS1/table.biom \\\n",
    "  --type 'FeatureTable[Frequency]' \\\n",
    "  --input-format BIOMV210Format \\\n",
    "  --output-path dada/ITS1/table.qza\n",
    "\n",
    "# add md5 hashes to sequence headers\n",
    "repseqs = list(SeqIO.parse('dada/ITS1/rep-seqs.fasta','fasta'))\n",
    "for r in repseqs:\n",
    "    r.id = hashlib.md5(str(r.seq).encode()).hexdigest()\n",
    "    r.description = ''\n",
    "SeqIO.write(repseqs,'temp1','fasta')\n",
    "\n",
    "# convert fasta to qza\n",
    "!qiime tools import \\\n",
    "  --input-path temp1 \\\n",
    "  --output-path dada/ITS1/rep-seqs.qza \\\n",
    "  --type 'FeatureData[Sequence]'\n",
    "\n",
    "!rm temp1\n",
    "\n",
    "# add md5 hashes to taxonomy table\n",
    "with open('temp.tsv','wt') as hndl:\n",
    "    first = True\n",
    "    for l in open('dada/ITS1/taxa.csv'):\n",
    "        if l.startswith('\"\"'):\n",
    "            continue\n",
    "        if first:\n",
    "            first=False\n",
    "        else:\n",
    "            hndl.write('\\n')\n",
    "        parts = l.rstrip().replace('\"','').split(',')\n",
    "        hndl.write(hashlib.md5(str(parts.pop(0)).encode()).hexdigest())\n",
    "        hndl.write('\\t')\n",
    "        hndl.write(';'.join(parts))\n",
    "        \n",
    "# convert taxonomy table to qza\n",
    "!qiime tools import \\\n",
    "   --type FeatureData[Taxonomy] \\\n",
    "   --input-path temp.tsv \\\n",
    "   --output-path dada/ITS1/taxonomy.qza \\\n",
    "   --input-format HeaderlessTSVTaxonomyFormat\n",
    "\n",
    "!rm temp.tsv\n",
    "!rm -r temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/ITS1/aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/ITS1/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: dada/ITS1/unrooted-tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: dada/ITS1/rooted-tree.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences dada/ITS1/rep-seqs.qza \\\n",
    "  --o-alignment dada/ITS1/aligned-rep-seqs.qza \\\n",
    "  --o-masked-alignment dada/ITS1/masked-aligned-rep-seqs.qza \\\n",
    "  --o-tree dada/ITS1/unrooted-tree.qza \\\n",
    "  --o-rooted-tree dada/ITS1/rooted-tree.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ITS1 metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('ITS1_metadata.tsv','wt') as hndl:\n",
    "    header = True\n",
    "    for l in open('16SrRNA_metadata.tsv'):\n",
    "        if header == True:\n",
    "            hndl.write(l)\n",
    "            header = False\n",
    "            continue\n",
    "        hndl.write('ITS-'+l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
