{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from gzip import open as gopen\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cut and merge the 16rRNA reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R1_count</th>\n",
       "      <th>R2_count</th>\n",
       "      <th>R1_mode_length</th>\n",
       "      <th>R1_count_at_mode_length</th>\n",
       "      <th>R2_mode_length</th>\n",
       "      <th>R2_count_at_mode_length</th>\n",
       "      <th>Mode_overlap</th>\n",
       "      <th>Count_at_mode_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AS3033</th>\n",
       "      <td>11047</td>\n",
       "      <td>11615</td>\n",
       "      <td>130</td>\n",
       "      <td>11047</td>\n",
       "      <td>130</td>\n",
       "      <td>10643</td>\n",
       "      <td>8</td>\n",
       "      <td>8153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3074</th>\n",
       "      <td>4145</td>\n",
       "      <td>5154</td>\n",
       "      <td>130</td>\n",
       "      <td>4145</td>\n",
       "      <td>130</td>\n",
       "      <td>4008</td>\n",
       "      <td>8</td>\n",
       "      <td>3019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3053</th>\n",
       "      <td>61</td>\n",
       "      <td>121</td>\n",
       "      <td>130</td>\n",
       "      <td>61</td>\n",
       "      <td>130</td>\n",
       "      <td>58</td>\n",
       "      <td>8</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3026</th>\n",
       "      <td>1317</td>\n",
       "      <td>1633</td>\n",
       "      <td>130</td>\n",
       "      <td>1317</td>\n",
       "      <td>130</td>\n",
       "      <td>1271</td>\n",
       "      <td>8</td>\n",
       "      <td>973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3057</th>\n",
       "      <td>26070</td>\n",
       "      <td>27878</td>\n",
       "      <td>130</td>\n",
       "      <td>26070</td>\n",
       "      <td>130</td>\n",
       "      <td>25483</td>\n",
       "      <td>8</td>\n",
       "      <td>20352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3039</th>\n",
       "      <td>9160</td>\n",
       "      <td>10175</td>\n",
       "      <td>130</td>\n",
       "      <td>9160</td>\n",
       "      <td>130</td>\n",
       "      <td>8943</td>\n",
       "      <td>8</td>\n",
       "      <td>6605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3007</th>\n",
       "      <td>11748</td>\n",
       "      <td>12539</td>\n",
       "      <td>130</td>\n",
       "      <td>11748</td>\n",
       "      <td>130</td>\n",
       "      <td>11263</td>\n",
       "      <td>8</td>\n",
       "      <td>8633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3009</th>\n",
       "      <td>15144</td>\n",
       "      <td>15669</td>\n",
       "      <td>130</td>\n",
       "      <td>15144</td>\n",
       "      <td>130</td>\n",
       "      <td>14737</td>\n",
       "      <td>8</td>\n",
       "      <td>11426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3023</th>\n",
       "      <td>30699</td>\n",
       "      <td>32463</td>\n",
       "      <td>130</td>\n",
       "      <td>30699</td>\n",
       "      <td>130</td>\n",
       "      <td>30041</td>\n",
       "      <td>8</td>\n",
       "      <td>24678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS3047</th>\n",
       "      <td>29918</td>\n",
       "      <td>31591</td>\n",
       "      <td>130</td>\n",
       "      <td>29918</td>\n",
       "      <td>130</td>\n",
       "      <td>29292</td>\n",
       "      <td>8</td>\n",
       "      <td>23692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       R1_count R2_count R1_mode_length R1_count_at_mode_length  \\\n",
       "AS3033    11047    11615            130                   11047   \n",
       "AS3074     4145     5154            130                    4145   \n",
       "AS3053       61      121            130                      61   \n",
       "AS3026     1317     1633            130                    1317   \n",
       "AS3057    26070    27878            130                   26070   \n",
       "...         ...      ...            ...                     ...   \n",
       "AS3039     9160    10175            130                    9160   \n",
       "AS3007    11748    12539            130                   11748   \n",
       "AS3009    15144    15669            130                   15144   \n",
       "AS3023    30699    32463            130                   30699   \n",
       "AS3047    29918    31591            130                   29918   \n",
       "\n",
       "       R2_mode_length R2_count_at_mode_length Mode_overlap  \\\n",
       "AS3033            130                   10643            8   \n",
       "AS3074            130                    4008            8   \n",
       "AS3053            130                      58            8   \n",
       "AS3026            130                    1271            8   \n",
       "AS3057            130                   25483            8   \n",
       "...               ...                     ...          ...   \n",
       "AS3039            130                    8943            8   \n",
       "AS3007            130                   11263            8   \n",
       "AS3009            130                   14737            8   \n",
       "AS3023            130                   30041            8   \n",
       "AS3047            130                   29292            8   \n",
       "\n",
       "       Count_at_mode_overlap  \n",
       "AS3033                  8153  \n",
       "AS3074                  3019  \n",
       "AS3053                    36  \n",
       "AS3026                   973  \n",
       "AS3057                 20352  \n",
       "...                      ...  \n",
       "AS3039                  6605  \n",
       "AS3007                  8633  \n",
       "AS3009                 11426  \n",
       "AS3023                 24678  \n",
       "AS3047                 23692  \n",
       "\n",
       "[74 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!mkdir -p cut_reads/16SrRNA\n",
    "\n",
    "for f in glob.glob('raw_data/16SrRNA/AS*_S*_L001_R*_001.fastq.gz'):\n",
    "    outf = 'cut_reads/16SrRNA/%s' % f.split('/')[-1]\n",
    "    smpl = f.split('/')[-1].split('_')[0]\n",
    "    records = SeqIO.parse(gopen(f,'rt'),'fastq')\n",
    "    with gopen(outf,'wt') as hndl:\n",
    "        for r in records:\n",
    "            \n",
    "            # Bad sequences have monomorphic ends.\n",
    "            # I skip sequences that were close to monomorphic in the last\n",
    "            # 40 base pairs\n",
    "            counts = Counter(str(r[110:150].seq).upper())\n",
    "            proportions = {i: counts[i]/sum(counts.values()) for i in counts.keys()}\n",
    "            if any([proportions[i] > 0.5 for i in proportions.keys()]):\n",
    "                continue\n",
    "                \n",
    "            hndl.write(r[20:150].format('fastq'))\n",
    "\n",
    "!mkdir -p merged/16SrRNA\n",
    "\n",
    "df = pd.DataFrame(columns=['R1_count','R2_count',\n",
    "                           'R1_mode_length','R1_count_at_mode_length',\n",
    "                           'R2_mode_length','R2_count_at_mode_length',\n",
    "                           'Mode_overlap','Count_at_mode_overlap'])\n",
    "\n",
    "for f1 in glob.glob('cut_reads/16SrRNA/AS*_S*_L001_R1_001.fastq.gz'):\n",
    "    f2 = f1.replace('_R1_','_R2_')\n",
    "    \n",
    "    smpl = f1.split('/')[-1].split('_')[0]\n",
    "    \n",
    "    merged = gopen('merged/16SrRNA/%s' % f1.split('/')[-1], 'wt')\n",
    "\n",
    "    r1s = list(SeqIO.parse(gopen(f1,'rt'),'fastq'))\n",
    "    r2s = SeqIO.to_dict(SeqIO.parse(gopen(f2,'rt'),'fastq'))\n",
    "    \n",
    "    count_r1 = len(r1s)\n",
    "    count_r2 = len(r2s)\n",
    "    \n",
    "    overlapps = []\n",
    "    r1_lengths= []\n",
    "    r2_lengths= []\n",
    "    \n",
    "    for r1 in r1s:\n",
    "        r1seq = str(r1.seq)\n",
    "        r1_lengths.append(len(r1))\n",
    "        try:\n",
    "            r2seq = str(r2s[r1.id].seq.reverse_complement())\n",
    "        except:\n",
    "            continue\n",
    "        r2_lengths.append(len(r2seq))\n",
    "        over_len = 30\n",
    "        r1_window = r1seq[-1*over_len:]\n",
    "        r2_window = r2seq[:over_len]\n",
    "        while over_len > 0 and r1_window != r2_window:\n",
    "            over_len -= 1\n",
    "            r1_window = r1seq[-1*over_len:]\n",
    "            r2_window = r2seq[:over_len]\n",
    "        overlapps.append(over_len)\n",
    "        if over_len > 5:\n",
    "            r2 = r2s[r1.id]\n",
    "            m = r1 + r2.reverse_complement()[over_len:]\n",
    "            m.id = r1.id\n",
    "            m.description = r1.description\n",
    "            merged.write(m.format('fastq'))\n",
    "    merged.close()        \n",
    "    if len(overlapps) == 0:\n",
    "        print( smpl)\n",
    "            \n",
    "    ovlp_mode, ovlp_count_at_mode = sorted(Counter(overlapps).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    r1len_mode, r1len_count_at_mode = sorted(Counter(r1_lengths).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    r2len_mode, r2len_count_at_mode = sorted(Counter(r2_lengths).items(), key=lambda k: k[1], reverse=True)[0]\n",
    "    df.loc[smpl] = [count_r1,count_r2,r1len_mode,r1len_count_at_mode,r2len_mode,r2len_count_at_mode,ovlp_mode,ovlp_count_at_mode]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('merged_16SrRNA_read_stas.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f3456265278>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXsAAAEHCAYAAABP3uaxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAS20lEQVR4nO3dfbRldV3H8fdHBh9Q8omrkjgOmlEslyneQMOUQIrMfChbgmlg2qyWUvS0DGItxdWTT9mTLW1KxAqQFAy1VFBAQnF0gAEGBwQUETVmzEorV4p8+2PvK5frvXfunLPPvYf5vV9r3XX2+Z299+97f3Pnc/fd5+zfTlUhSdqz3WutC5AkTZ5hL0kNMOwlqQGGvSQ1wLCXpAasW83O9ttvv9qwYcNqdilJ93hXXHHFV6tqZpx9rGrYb9iwgS1btqxml5J0j5fkC+Puw9M4ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQGGvSQ1wLCXpAYY9pLUgFW9gla756zNt65ovRcdtn7ClUi6p/PIXpIaYNhLUgMMe0lqgGEvSQ0w7CWpAYa9JDXAsJekBhj2ktQAw16SGrDLsE9yepIdSbYt8trvJKkk+02mPEnSEFZyZH8GcMzCxiSPAo4GVnZNvyRpzewy7KvqUuBri7z0p8CrgBq6KEnSsEY6Z5/kOcCXqurqgeuRJE3Abs96mWQf4FTgJ1e4/kZgI8D69c7OKElrYZQj+8cCBwJXJ7kFOAC4MskjFlu5qjZV1WxVzc7MzIxeqSRpZLt9ZF9V1wIPm3veB/5sVX11wLokSQNayUcvzwYuBw5KcluSl02+LEnSkHZ5ZF9Vx+3i9Q2DVSNJmgivoJWkBhj2ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQGGvSQ1wLCXpAYY9pLUAMNekhpg2EtSAwx7SWqAYS9JDTDsJakBhr0kNcCwl6QGGPaS1ADDXpIasJIbjp+eZEeSbfPa3pjk+iTXJHlvkgdNtkxJ0jhWcmR/BnDMgrYLgcdX1ROAzwKnDFyXJGlAuwz7qroU+NqCtguq6o7+6SeBAyZQmyRpIEOcs/9l4INLvZhkY5ItSbbs3LlzgO4kSbtrrLBPcipwB3DmUutU1aaqmq2q2ZmZmXG6kySNaN2oGyY5Hng2cFRV1XAlSZKGNlLYJzkG+F3gGVX1v8OWJEka2ko+enk2cDlwUJLbkrwMeAuwL3Bhkq1J3jbhOiVJY9jlkX1VHbdI89snUIskaUK8glaSGmDYS1IDDHtJaoBhL0kNMOwlqQGGvSQ1wLCXpAYY9pLUAMNekhpg2EtSAwx7SWqAYS9JDTDsJakBhr0kNcCwl6QGGPaS1ADDXpIaYNhLUgMMe0lqwEpuOH56kh1Jts1re0iSC5Pc2D8+eLJlSpLGsZIj+zOAYxa0nQx8tKoeB3y0fy5JmlK7DPuquhT42oLm5wLv7JffCTxv4LokSQMa9Zz9w6vqKwD948OWWjHJxiRbkmzZuXPniN1JksYx8Tdoq2pTVc1W1ezMzMyku5MkLWLUsL89yf4A/eOO4UqSJA1t1LB/H3B8v3w8cP4w5UiSJmElH708G7gcOCjJbUleBrwOODrJjcDR/XNJ0pRat6sVquq4JV46auBaJEkT4hW0ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQGGvSQ1wLCXpAbs8qIqDe+szbeudQmSGuORvSQ1wLCXpAYY9pLUAMNekhpg2EtSAwx7SWqAYS9JDTDsJakBhr0kNWCssE/ym0muS7ItydlJ7jtUYZKk4Ywc9kkeCfw6MFtVjwf2Ao4dqjBJ0nDGPY2zDrhfknXAPsCXxy9JkjS0kSdCq6ovJXkTcCvwTeCCqrpg4XpJNgIbAdavXz9qdyuePOxFh43ehyTtqcY5jfNg4LnAgcD3A/dP8uKF61XVpqqararZmZmZ0SuVJI1snNM4zwQ+X1U7q+rbwHnAjw1TliRpSOOE/a3AU5LskyTAUcD2YcqSJA1p5LCvqs3Ae4ArgWv7fW0aqC5J0oDGulNVVb0GeM1AtUiSJsQraCWpAYa9JDXAsJekBhj2ktQAw16SGmDYS1IDDHtJaoBhL0kNGOuiKt3dSmfmlKTV5pG9JDXAsJekBhj2ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQGGvSQ1wLCXpAaMFfZJHpTkPUmuT7I9yVOHKkySNJxx58b5c+BDVfWCJPcG9hmgJknSwEYO+yTfBzwdOAGgqr4FfGuYsiRJQxrnNM5jgJ3AO5JcleRvk9x/4UpJNibZkmTLzp07x+hOkjSqccJ+HXAI8NaqehLwP8DJC1eqqk1VNVtVszMzM2N0J0ka1ThhfxtwW1Vt7p+/hy78JUlTZuSwr6p/A76Y5KC+6SjgM4NUJUka1Lifxvk14Mz+kzifA146fkmSpKGNFfZVtRWYHagWSdKEeAWtJDXAsJekBhj2ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQHjXkGrPdBZm29d8bovOmz9BCuRNBSP7CWpAYa9JDXAsJekBhj2ktQAw16SGmDYS1IDDHtJaoBhL0kNMOwlqQFjh32SvZJcleQDQxQkSRreEEf2JwHbB9iPJGlCxgr7JAcAPwP87TDlSJImYdwj+z8DXgXcOUAtkqQJGXnWyyTPBnZU1RVJjlhmvY3ARoD1650hcRJWOkulM1RK7RrnyP5w4DlJbgHeBRyZ5B8WrlRVm6pqtqpmZ2ZmxuhOkjSqkcO+qk6pqgOqagNwLHBRVb14sMokSYPxc/aS1IBB7lRVVZcAlwyxL0nS8Dyyl6QGGPaS1ADDXpIaYNhLUgMMe0lqgGEvSQ0w7CWpAYa9JDVgkIuq7olWOnkY7DkTiO3O9yxpz+KRvSQ1wLCXpAYY9pLUAMNekhpg2EtSAwx7SWqAYS9JDTDsJakBhr0kNcCwl6QGjBz2SR6V5OIk25Ncl+SkIQuTJA1nnLlx7gB+u6quTLIvcEWSC6vqMwPVJkkayMhH9lX1laq6sl/+BrAdeORQhUmShjPIrJdJNgBPAjYv8tpGYCPA+vWTnz3SmR01ipX+3OwpM6BCm99zy8Z+gzbJA4Bzgd+oqq8vfL2qNlXVbFXNzszMjNudJGkEY4V9kr3pgv7MqjpvmJIkSUMb59M4Ad4ObK+qNw9XkiRpaOMc2R8OvAQ4MsnW/utZA9UlSRrQyG/QVtVlQAasRZI0IV5BK0kNMOwlqQGGvSQ1wLCXpAYY9pLUAMNekhpg2EtSAwx7SWrAILNe7umcSXNpQ4/NSmdYvCfM2Njiz82e8j0P/XO4O/ucFI/sJakBhr0kNcCwl6QGGPaS1ADDXpIaYNhLUgMMe0lqgGEvSQ0w7CWpAYa9JDVgrLBPckySG5LclOTkoYqSJA1r5LBPshfwV8BPAwcDxyU5eKjCJEnDGefI/lDgpqr6XFV9C3gX8NxhypIkDSlVNdqGyQuAY6rq5f3zlwCHVdWJC9bbCGzsnx4E3DBirfsBXx1x20mb5tpguuuzttFNc33WNrrF6nt0Vc2Ms9NxpjjOIm3f85ujqjYBm8bop+ss2VJVs+PuZxKmuTaY7vqsbXTTXJ+1jW5S9Y1zGuc24FHznh8AfHm8ciRJkzBO2H8aeFySA5PcGzgWeN8wZUmShjTyaZyquiPJicCHgb2A06vqusEq+15jnwqaoGmuDaa7Pmsb3TTXZ22jm0h9I79BK0m65/AKWklqgGEvSQ2Y+rBfqykZktyS5NokW5Ns6dsekuTCJDf2jw+et/4pfY03JPmpee1P7vdzU5K/SLLYR1ZXUs/pSXYk2TavbbB6ktwnyTl9++YkG8as7bQkX+rHb2uSZ61RbY9KcnGS7UmuS3LStIzdMrVNy9jdN8mnklzd1/faKRq7pWqbirHrt98ryVVJPjAV41ZVU/tF98bvzcBjgHsDVwMHr1LftwD7LWh7A3Byv3wy8Pp++eC+tvsAB/Y179W/9ingqXTXJXwQ+OkR63k6cAiwbRL1AK8A3tYvHwucM2ZtpwG/s8i6q13b/sAh/fK+wGf7GtZ87JapbVrGLsAD+uW9gc3AU6Zk7JaqbSrGrt/mt4CzgA9Mw//XiYfmOF/9N/nhec9PAU5Zpb5v4XvD/gZg/355f+CGxeqi+4TSU/t1rp/Xfhzw12PUtIG7B+pg9cyt0y+vo7uCL2PUttR/ulWvbUH/5wNHT9PYLVLb1I0dsA9wJXDYtI3dgtqmYuzorjv6KHAkd4X9mo7btJ/GeSTwxXnPb+vbVkMBFyS5It2UDwAPr6qvAPSPD9tFnY/slxe2D2XIer67TVXdAfwX8NAx6zsxyTXpTvPM/cm6ZrX1f+o+ie4ocKrGbkFtMCVj15+K2ArsAC6sqqkZuyVqg+kYuz8DXgXcOa9tTcdt2sN+RVMyTMjhVXUI3ayer0zy9GXWXarOtap/lHqGrvWtwGOBJwJfAf5kLWtL8gDgXOA3qurry6262vUtUtvUjF1Vfaeqnkh3pHpokscvs/qq1rdEbWs+dkmeDeyoqit2te5q1jbtYb9mUzJU1Zf7xx3Ae+lm+bw9yf4A/eOOXdR5W7+8sH0oQ9bz3W2SrAMeCHxt1MKq6vb+P+OdwN/Qjd+a1JZkb7owPbOqzuubp2LsFqttmsZuTlX9J3AJcAxTMnaL1TYlY3c48Jwkt9DNBnxkkn9gjcdt2sN+TaZkSHL/JPvOLQM/CWzr+z6+X+14unOs9O3H9u+QHwg8DvhU/6faN5I8pX8X/ZfmbTOEIeuZv68XABdVf0JwFHM/1L3n043fqtfW7+vtwPaqevO8l9Z87JaqbYrGbibJg/rl+wHPBK5nOsZu0dqmYeyq6pSqOqCqNtBl1kVV9WLWetx2542QtfgCnkX3KYWbgVNXqc/H0L07fjVw3Vy/dOfEPgrc2D8+ZN42p/Y13sC8T9wAs3Q/cDcDb2H0NxbPpvuz9Nt0v9VfNmQ9wH2BdwM30X0C4DFj1vb3wLXANf0P5v5rVNvT6P68vQbY2n89axrGbpnapmXsngBc1dexDXj10P8Pxhi7pWqbirGbt+8juOsN2jUdN6dLkKQGTPtpHEnSAAx7SWqAYS9JDTDsJakBhr0kNcCwl6QGGPZaVpJHJHlXkpuTfCbJvyT5wQH3f0SSH1vtbUfs74Qkb1mt/nal//4/sNZ16J7BsNeS+qv23gtcUlWPraqDgd8DHj5gN0cAowb2ONveo/WXyEsrZthrOT8BfLuq3jbXUFVbgcuSvDHJtnQ3VnghfO+RZpK3JDmhX74lyWuTXNlv80P9TI+/CvxmuhtN/PhiRST52XQ3aLgqyUeSPHw3tj0jyVvT3STkc0mekW42xO1Jzpi33nF9XduSvH5e+0uTfDbJx+jmPJlrn0lybpJP91+Hs4R0N634p3QzMX4yyROS3KsfkwfNW++m/ntbdN/pbsyxKckFwN8t6OPQJJ/ox+gTSQ7q209Icn6SD6W7McZrlqpTezaPDrScxwOLzdz3c3SzCv4IsB/w6SSXrmB/X62qQ5K8gm7O8ZcneRvw31X1pmW2uwx4SlVVkpcDr6qq317htgAPpptX/DnA++lC++V93U+km5Dq9cCTgf+gm9r6eXTTDb+2b/8v4GK6S/QB/hz406q6LMl6uvnFf3iJ/l8LXFVVz0tyJPB3VfXEJOfTzd/yjiSHAbdU1e1Jzlpm308GnlZV30xyxLw+rgeeXlV3JHkm8EfAz/evHUr3b/m//ff8z1W1ZRdjpj2MYa9RPA04u6q+QzeT38eAHwWWmzoYYG7GySvofmGs1AHAOf0kV/cGPr+b9b6//0VxLXB7VV0LkOQ6upuuPJruVNXOvv1MurtvsaD9HGDu/YpnAgfnrrtMfl+SfavqG4v0/zT64K2qi5I8NMkDgXOAVwPvoL/b0HL77pffV1XfXKSPBwLvTPI4uvl29p732oVV9e/993BeX49h3xjDXsu5jm5GvYWWuo/uHdz91OB9F7z+f/3jd9i9n72/BN5cVe/rj2ZP241t5/d757zluefr6OpeylKTR92L7k5BiwXvQkvNPX458ANJZoDnAX+w3L778P+fJfr4feDiqnp+f4rrkgV9LexbjfGcvZZzEXCfJL8y15DkR+lOdbww3Z2CZuiOgj8FfIHuiPQ+/ZHrUSvo4xt0919dzgOBL/XLx89rX8m2K7EZeEaS/ZLsRXf7t4/17Uf0R+J7A78wb5sLgBPnnvSng5ZyKfCL/XpH0J3O+np1sxC+F3gz3TTH/z7CvufMH6MTFrx2dP++wf3ofql8fAX70x7GsNeS+jB6Pl1Y3Nyf9jiN7ibK19BNAX0R3Tn0f6uqLwL/2L92Jned317O+4HnL/cma9/nu5P8K929Nndn212qbt7wU+jOyV8NXFlV5/ftp9EdgX+E7j6nc34dmO3fdP0M3ZvFSzltbl3gddz9F9Y5wIu56xTO7u57zhuAP07ycWCvBa9dRjf171bgXM/Xt8kpjqU9WLpPQ81W1Ym7Wld7No/sJakBHtlraiQ5lbufFwd4d1X94SS3HUqSlwInLWj+eFW9crVqkJZi2EtSAzyNI0kNMOwlqQGGvSQ1wLCXpAb8P9e8HbM/B+jbAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "\n",
    "df = pd.read_csv('merged_16SrRNA_read_stas.tsv', sep='\\t',index_col=0)\n",
    "sns.distplot(df.Count_at_mode_overlap,kde=False,bins=30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import merged reads to qiime2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/amir/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n",
      "\u001b[32mImported merged/16SrRNA as CasavaOneEightSingleLanePerSampleDirFmt to merged/16SrRNA/demux.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime tools import \\\n",
    "  --type 'SampleData[SequencesWithQuality]' \\\n",
    "  --input-path merged/16SrRNA \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path merged/16SrRNA/demux.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run DADA2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘dada/16SrRNA’: File exists\n",
      "/home/amir/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n",
      "\u001b[32mSaved FeatureTable[Frequency] to: dada/16SrRNA/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: dada/16SrRNA/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: dada/16SrRNA/stats.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!mkdir dada/16SrRNA\n",
    "!qiime dada2 denoise-single \\\n",
    "  --i-demultiplexed-seqs merged/16SrRNA/demux.qza \\\n",
    "  --p-trim-left 0 \\\n",
    "  --p-trunc-len 250 \\\n",
    "  --p-max-ee 5.0 \\\n",
    "  --o-representative-sequences dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-table dada/16SrRNA/table.qza \\\n",
    "  --o-denoising-stats dada/16SrRNA/stats.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/amir/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n",
      "\u001b[32mSaved Visualization to: dada/16SrRNA/stats.qzv\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file dada/16SrRNA/stats.qza \\\n",
    "  --o-visualization dada/16SrRNA/stats.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign taxonomy to ASVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-27 08:35:31--  https://data.qiime2.org/2020.8/common/silva-138-99-515-806-nb-classifier.qza\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/common/silva-138-99-515-806-nb-classifier.qza [following]\n",
      "--2020-12-27 08:35:32--  https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/common/silva-138-99-515-806-nb-classifier.qza\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.218.209.24\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.218.209.24|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 153517385 (146M) [binary/octet-stream]\n",
      "Saving to: ‘DB/silva-138-99-515-806-nb-classifier.qza’\n",
      "\n",
      "DB/silva-138-99-515 100%[===================>] 146.41M  4.06MB/s    in 51s     \n",
      "\n",
      "2020-12-27 08:36:24 (2.86 MB/s) - ‘DB/silva-138-99-515-806-nb-classifier.qza’ saved [153517385/153517385]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://data.qiime2.org/2020.8/common/silva-138-99-515-806-nb-classifier.qza -O DB/silva-138-99-515-806-nb-classifier.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/amir/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n",
      "\u001b[32mSaved FeatureData[Taxonomy] to: dada/16SrRNA/taxonomy.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier classify-sklearn \\\n",
    "  --i-classifier DB/silva-138-99-515-806-nb-classifier.qza \\\n",
    "  --i-reads dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-classification dada/16SrRNA/taxonomy.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make ASV tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/amir/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/util/_testing.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as pdt\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/16SrRNA/aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: dada/16SrRNA/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: dada/16SrRNA/unrooted-tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: dada/16SrRNA/rooted-tree.qza\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences dada/16SrRNA/rep-seqs.qza \\\n",
    "  --o-alignment dada/16SrRNA/aligned-rep-seqs.qza \\\n",
    "  --o-masked-alignment dada/16SrRNA/masked-aligned-rep-seqs.qza \\\n",
    "  --o-tree dada/16SrRNA/unrooted-tree.qza \\\n",
    "  --o-rooted-tree dada/16SrRNA/rooted-tree.qza"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
